### PR TITLE
[SPARK-47495][CORE] Fix primary resource jar added to spark.jars twice under k8s cluster mode

### PR DESCRIPTION
**Pull request by leletan**
### What changes were proposed in this pull request?

In `SparkSubmit`, for `isKubernetesClusterModeDriver` code path, stop appending primary resource to `spark.jars` to avoid duplicating the primary resource jar in `spark.jars`.

#### Context:

To submit spark jobs to Kubernetes under cluster mode, the spark-submit will be called twice. The first time SparkSubmit will run under k8s cluster mode, it will append primary resource to `spark.jars` and call `KubernetesClientApplication::start`  to create a driver pod. The driver pod will run spark-submit again with the updated configurations (with the same application jar  but that jar will also be in the `spark.jars`). This time the SparkSubmit will run under client mode with `spark.kubernetes.submitInDriver`  as `true`. Under this mode, all the jars in `spark.jars` will be downloaded to driver and jars' urls will be replaced by the driver local paths. Later SparkSubmit will append primary resource to `spark.jars` again. So in this case, `spark.jars` will have 2 paths of duplicate copies of primary resource, one with the original url user submit with, the other with the driver local file path. Later when driver starts the `SparkContext` it will copy all the `spark.jars` to `spark.app.initial.jar.urls`, and replace the driver local jars paths in `spark.app.initial.jar.urls` with driver file service paths, with which the executor can download those driver local jars. 

#### Issues:
The executor will download 2 duplicate copies of primary resource, one with the original url user submit with, the other with the driver local file path, which leads to resource waste.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test added.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by leletan** [comment]
retest this please

**Comment by dbtsai** [comment]
cc @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Ack, @dbtsai .

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR, @leletan .

This PR seems to mix two independent themes. Please create another JIRA for the following. We can proceed that first.
> Stop appending primary resource to spark.jars to avoid duplicating the primary resource jar in spark.jars.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1458,6 +1458,18 @@ package object config {
...
       .createWithDefault(1.5)
 
+  private[spark] val KUBERNETES_AVOID_JAR_DOWNLOAD_SCHEMES =
+    ConfigBuilder("spark.kubernetes.jars.avoidDownloadSchemes")
+      .doc("Comma-separated list of schemes for which jars will not be downloaded to the " +
+        "driver local disk prior to be distributed to executors, only for kubernetes deployment. " +
+        "For use in cases when the jars are big and executor counts are high, " +
+        "concurrent download causes network saturation and timeouts. " +
+        "Wildcard '*' is denoted to not downloading jars for any the schemes.")
+      .version("2.3.0")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:1468]: This should be `4.0.0`.

Comment (leletan) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1468]: Good catch.

Comment (leletan) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala]: Will move this to another JIRA & PR.

**Comment by leletan** [comment]
@dongjoon-hyun Updated the PR and associated it with the new JIRA https://issues.apache.org/jira/browse/SPARK-47495. Please let me know if this looks good to you. Thanks!

**Comment by mridulm** [comment]
+CC @zhouyejoe

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @leletan .

I'll resume the review this weekend.

**Comment by dongjoon-hyun** [comment]
I just assigned this to me in order not to forget. It doesn't block any community reviews.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM for Apache Spark 4.0.0.

Thank you, @leletan , @dbtsai , @mridulm .

Merged to master.

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala [hunk 2]
```diff
@@ -504,6 +504,25 @@ class SparkSubmitSuite
     }
   }
 
+  test("SPARK-47475: Not to add primary resource to jars again" +
```
Comment (dongjoon-hyun) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:507]: Oh, this JIRA ID is wrong. We need to have SPARK-47495 like the PR title.

**Comment by dongjoon-hyun** [comment]
Merged to master because the last commit only changes JIRA ID in the test case name.

**Comment by dongjoon-hyun** [comment]
Welcome to the Apache Spark community, @leletan . 

I added you to the Apache Spark contributor group (in JIRA) and assigned SPARK-47495 to you.

Congratulations for your first commit, @leletan .

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala (hunk 2 above)
Comment (leletan) [reply core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala]: Good catch!!!
Thanks for fixing this!

**Pull request closed without merging**